src/net/client/factory.rs

```diff
@@ -0,0 +1,22 @@
+//! Trait for connection factories
+
+#![warn(missing_docs)]
+#![warn(clippy::missing_docs_in_private_items)]
+
+use std::boxed::Box;
+use std::future::Future;
+use std::pin::Pin;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+/// This trait is for creating new network connections.
+///
+/// The IO type is the type of the resulting connection object.
+pub trait ConnFactory<IO: AsyncRead + AsyncWrite + Send + Unpin> {
+    /// The next method is an asynchronous function that returns a
+    /// new connection.
+    ///
+    /// This method is equivalent to async fn next(&self) -> Result<IO, std::io::Error>;
+    fn next(
+        &self,
+    ) -> Pin<Box<dyn Future<Output = Result<IO, std::io::Error>> + Send + '_>>;
```

Comment:
I replaced the Pin<Box< with an associated type. I don't know how to get rid of the Pin<Box< in the implementation. So it doesn't seem to have any net result.

Reply:
Async trait functions will be stabilized real soon (next release?), so maybe we should just switch to that, use the feature flag for now and only release one it is stable?

Reply:
It would be annoying to roll back if we need to release before async trait functions are released. And it should be a small amount of work when they are released. So maybe for now we can concentrate on other parts. The factory is implemented in only two places and I guess for using it, it doesn't really matter.